### ISSUE TITLE
wicket should be able to update the scrimlet in a single sidecar environment

### ISSUE THREAD

**Issue by jclulow** [oxidecomputer/omicron#4905]
At present, it does not appear to be possible to use **wicket** to perform a **mupdate** of an entire environment when there is only one sidecar: **wicket** cannot coordinate update of the scrimlet upon which it is executing, so the user must perform some out-of-band manual update of at least the host OS image there.

In the past, a hard limit on improving this situation has been that while the scrimlet is online (in the A0 state) it has not been possible to update the host OS ROM image over the management network.  Taking the scrimlet down to A2 means that wicket (and indeed all host OS software!) is obviously not executing any longer.

As of oxidecomputer/hubris#1575 and oxidecomputer/hubris#1172 it is now actually possible to update the host OS ROM while a system is in A0, provided it is not currently booting.  The first IPCC message sent from the host OS to the SP causes the SP to detach the host OS ROM, making it available for updates again (in A0!) as it would be in the A2 state.

There is still the question of what to do about the fact that mupdate is driven by a user with an SSH connection to the switch zone, but I suspect what we'd want to do is roughly:

* line up all of the ducks on other sleds that need updating, then get the M.2 disk image deployed on the scrimlet
* get the host OS ROM updated on the scrimlet over the management network
* tell the user something to the effect of "it's now safe to turn off your computer" and potentially gracefully end the SSH connection and kick off a reboot
* the user would dutifully wait a number of minutes and then be able to SSH back in to the freshly updated and rebooted switch zone

**Comment by jgallagher** [comment]
Agreed that I don't think there are any hard blockers left after the hubris changes. A couple notes on how this will be different from the way wicket currently drives updates:

* We would neither need to, want to, nor be _able_ to use installinator to drive the update to ourself the way it drives updates to other sleds. Installinator exists to allow recovery of a completely-software-dead (e.g., empty M.2s) sled, but is overkill if we assume the scrimlet where wicket is running has a functional sled-agent (which it at least did at one point, if the switch zone is running).
* Instead, we'd want wicketd to talk to the sled-agent in its own gz, and send it the control plane artifacts and the real host phase2. (sled-agent APIs for writing these exist today; the real host phase2 is designed for exactly this, but I think the API for updating control plane artifacts was a demo/placeholder so might need some work.

We'd then proceed with the rest of the steps you outlined (send the phase1 over the management network, etc)

**Comment by jgallagher** [comment]
A few random thoughts from poking at this a tiny bit:

* When we mupdate we also update the RoT and SP. Self-updating those seems a _little_ tricky. Today, we can reset the RoT without affecting anything else (but that could change in the future, where an RoT reset forces a reset of everything down the line). Resetting the SP resets the host, though. So I think today we could order things as "update and reset the RoT, stage the update to the host, update and reset the SP (which also resets the host)", but that'll break once resetting the RoT resets the SP/host.
* Updating the SP of the scrimlet's own sidecar is similarly tricky. wicket currently blocks this too, because I believe if the switch resets we need to reset the scrimlet.

**Comment by jclulow** [comment]
I think you could certainly get a lot of mileage out of an incremental improvement that only deals with the host OS!

As for the other thing, if we can't reset a sidecar without also resetting the scrimlet manually after that, I think we need to get a bug filed for whatever the reason is there.

**Comment by Nieuwejaar** [comment]
If `dendrite` crashes, `smf` will attempt to restart it.  If the switch comes back quickly enough, then the restarted `dendrite` should be able to connect to the `tofino` and start running as normal.

If the switch is offline for too long, `dendrite` will keep crashing on startup, and `smf` will eventually stop trying to restart it.  At that point a `svcadm clear dendrite` should be enough to get things rolling again.

So, I don't believe that restarting the switch zone is strictly necessary, but it might be simpler and more reliable than trying to determine where we are in this recovery process.
